### PR TITLE
Feature/add dcs scan dump

### PR DESCRIPTION
**Pull request by itayx**
Add "Dwell time = 0" feature to allow for the DCS task to request a scan dump.
A scan dump consists of the BWL "dumping" all available cached results
#818

### CODE REVIEW

File: common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp
Comment (vitalybu) [inline common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:163]: Shouldn't this method return `true` for the dummy bwl?
Otherwise the DCS logic will always fail on Linux.
Maybe it's better to return true here and provide an empty result-set.

File: common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp:398]: Same comment as for the dummy bwl

File: agent/src/beerocks/monitor/monitor_thread.cpp
Comment (vitalybu) [inline agent/src/beerocks/monitor/monitor_thread.cpp:1264]: I don't understand this comment

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (vitalybu) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This kind of code is used widely to forward messages between the different components.
Perhaps we should create a dedicated method for forwarding messages. Obviously it shouldn't be part of this PR.
@arnout @tomereli what do you guys think?

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:33]: Remove the extra newline

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]:
```suggestion
        LOG(ERROR) << "Invalid dwell_time = " << dwell_time_msec;
```

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:70]: A better error message would be:
```C++
LOG(ERROR) << "Unable to trigger dynamic channel scan without a pre-defined channels pool. Please define the requested channels for radio: " << m_radio_mac;
```

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:75]:
```suggestion
        LOG(ERROR) << "DCS supported channel list is too large: " << curr_channel_pool.size() << "/" << beerocks::message::SUPPORTED_CHANNELS_LENGTH;
```

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:125]: Should you first check that this flag is not already `true`?

File: common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp
Comment (itayx) [reply common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:163]: Other not implemented functions also return false,
I'm not sure if it's a good thing to break consistency

File: agent/src/beerocks/monitor/monitor_thread.cpp
Comment (itayx) [reply agent/src/beerocks/monitor/monitor_thread.cpp:1264]: This is how we "trick" the task into moving to the next current state
Incoming: ACTION_MONITOR_CHANNEL_DUMP_SCAN_RESULTS_REQUEST
Outgoing:
- ACTION_MONITOR_CHANNEL_DUMP_SCAN_RESULTS_RESPONSE
- ACTION_MONITOR_CHANNEL_SCAN_RESULTS_NOTIFICATION

Most requests only return a single response, but this message sends one response and one notification immediately

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:125]: You cannot reach this state if the flag is already `true`

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:33]: Removed

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: Changed

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:70]: Agree, Changed

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:75]: Changed

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Well, with the current tlvf, forwarding messages is not so simple, and in this case it even changes class (cACTION_CONTROL_CHANNEL_DUMP_SCAN_RESULTS_REQUEST -> cACTION_MONITOR_CHANNEL_DUMP_SCAN_RESULTS_REQUEST). So there are a few possibilities here.

- Ignore the issue. Many of the copies will disappear by themselves because we will use EasyMesh messages from the slave onwards instead of VS messages.
- Stop creating separate classes for each connection, and just use cACTION_CHANNEL_DUMP_SCAN_RESULTS_REQUEST; add a forwarding function to the slave.
- Make sure that all the details are in a substructure, so the forwarding is simply a copy of the substructure even when forwarding. We may have to add an assignment operator to tlvf to make sure this works.
- Allow all threads to connect to the bus directly, so no UDS between components is needed any more. This removes the need for forwarding entirely.

The latter has my personal preference as the long-term solution, but is possibly the most work. It also requires every platform to support a full transport layer, so we probably really need to replace zmq with something else then.

File: common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp
Comment (kantera800) [reply common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:163]: dummy mode currently doesn't have DCS  tests so its better to leave the false .

File: common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
Comment (kantera800) [reply common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp:398]: I think the false is correct since we don't have any DCS implementation in nl80211 bwl.

Comment (itayx) [reply common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp:398]: Just like Alex said, since there is no implementation I think returning false would be the correct way

File: agent/src/beerocks/monitor/monitor_thread.cpp
Comment (vitalybu) [reply agent/src/beerocks/monitor/monitor_thread.cpp:1264]: Why do we need a RESPONSE and a NOTIFICATION message?
Can't we use only a NOTIFICATION message?

When sending a response to a request, it'll contain the ID of the request.
When sending async notifications, the ID will remain 0.

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (vitalybu) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: I don't see the change

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (vitalybu) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]:
```suggestion
        std::cout << __func__ << ", invalid input: dwell_time(" << dwell_time << ") < 0"
```

File: controller/src/beerocks/master/db/db.cpp
Comment (vitalybu) [inline controller/src/beerocks/master/db/db.cpp]:
```suggestion
        LOG(ERROR) << "Invalid dwell time: " << dwell_time_msec << ". Only positive values are supported!";
```

File: agent/src/beerocks/monitor/monitor_thread.cpp
Comment (itayx) [reply agent/src/beerocks/monitor/monitor_thread.cpp:1264]: Since the response is part of the request-response flow and the notification has nothing to do it .
The notification is sent when the driver sends a `scan_triggered` NL event
But when using `dwell time = 0` the we never request a scan from the driver so the Nl event will never be sent

so to "trick" the FSM in the dcs' task to work as if a scan was triggered, we send the notification back manually while not actually triggering the bml channel scan

File: controller/src/beerocks/master/db/db.cpp
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: Changed

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (itayx) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: Changed

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml]: It's weird that you add the two new messages in the middle instead of at the end (i.e. the two new ones would be numbered 146 and 147).

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml]: Is this message even useful? I'd expect the reply to be either a cACTION_CONTROL_CHANNEL_SCAN_RESULTS_NOTIFICATION or something like cACTION_CONTROL_CHANNEL_SCAN_ABORT_NOTIFICATION.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml]: It would be nicer to make them more similar to the other message names, i.e. `CHANNEL_SCAN_DUMP_RESULTS_REQUEST`.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml]: The commit message that introduces this is fairly confusing. It speaks of "dwell time" but that's pretty unrelated to these new messages. IIUC, the idea is to use the a dwell time of 0 in the user interface (i.e. BML) to indicate that you want to use cached results instead of doing a scan. But these two new messages are just about using cached results. Therefore, I'd reword the commit message as follows:

```
 tlfv: Add dump results request and response

The dynamic channel scan will switch channels to scan for beacons
from neighboring APs. Sometimes, we may want to get scan results
without switching channels, e.g. cached results or results from scanning
the channel that the AP is operating on.

Add two new messages, CHANNEL_DUMP_SCAN_RESULTS_REQUEST
and CHANNEL_DUMP_SCAN_RESULTS_RESPONSE, to get scan results
without switching channels. The new messages are added both to controller
and monitor.
```

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp]: As I wrote: the response isn't used for anything, so there is no need to add it.

There is of course the symmetry of having a request/response for everything. However, the symmetry is already broken a bit through the different notification messages. And for sure, I don't think it warrants adding roughly 100 lines of code just for symmetry.

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (arnout) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: This should be `<= 0` IMO.

**Review by arnout — APPROVED** [review]
None of my comments are essential in practice. Since this anyway requires a manual rebase, I'm approving as is, and you can take the comments into account as you wish.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
Comment (itayx) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml]: I wanted to add them in relation to where they would be used in the scan flow
And since the `scan dump request` is an alternative version of the `scan trigger request` I thought it was best to add them after the scan trigger request (and response)

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml
Comment (itayx) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml]: You're right,
changed the commit message

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml
Comment (itayx) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml]: You're right the message isn't in use other then to align with the CMDU symmetry
It also has a side benefit when it comes to logs, as a confirmation that the request was sent correctly and that the dump request operation finished successfully
Note: This does not mean that the dump is successful, just that the request operation was

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (itayx) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: I think the best course would be to use the second option mentioned by @arnout, using a non-separate messages for forward-only message would reduce the code added, all that would be needed is a static function in the slave, this method should only be used if the messages do not have any additional function in the slave.

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (itayx) [reply controller/src/beerocks/master/son_master_thread.cpp]: I partly agree, while I do think that adding code for symmetry's sake is not a good reason, I have used the additional logs that the response prints to validate that the request has gone through and has performed correctly so I still see a use for the request/response symmetry, even if the only reason is logging

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: I understand why, but keep in mind that the only way to reach this code is if the dwell time `!= 0`
Changed to `<= 0` anyway

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (morantr) [inline controller/src/beerocks/master/son_master_thread.cpp:3387]: "... the response is only used as confirmation for the request".
I would expect to see the confirmation. Sending an event to the pending caller, or something similar.

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (morantr) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: For new code, better use `!share_ptr` than `shared_ptr == nullptr`.
You repeat it several times in this commit.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: This commit (67a5cb412e2f938b4334e7fc5fd9ed6dba300c10: beerocks_cli_bml: start_dcs_single_scan: remove dwell time 0 varifica…) headline is too long and exeeds 72 characters.

Also, the print would look like this:
"invalid input: dwell_time(-3) < 0. Only positive values are supported!"

I think it would look better like this:
"invalid input, dwell_time='-3', Only positive values are supported!
(nitpick)

File: controller/src/beerocks/master/db/db.cpp
Comment (morantr) [inline controller/src/beerocks/master/db/db.cpp]: There is no need to mention `(>=0)` since it is the definition of `positive number`.

**Review by morantr — APPROVED** [review]
Small comments. I approved and labeled with "don't merge" so you will have a chance to change it before it automatically merged.

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (itayx) [reply controller/src/beerocks/master/son_master_thread.cpp:3387]: The confirmation is due to the Log prints,
It's a way to print in the controller log that the action sent from the controller was successful

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp]: Changed

File: controller/src/beerocks/master/db/db.cpp
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: I wanted to remove the ambiguity with `positive number` and a `0` value

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (itayx) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: I agree it looks better without the `<0`
Changed

**Pull request merged**
